src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs

```diff
@@ -12,76 +11,91 @@ internal sealed partial class LocalRewriter
+                if (deconstruction.DeconstructMemberOpt == null)
+                {
+                    // tuple case
+                    AccessTupleFields(node, deconstruction, temps, stores, placeholders);
+                }
+                else
+                {
+                    CallDeconstruct(node, deconstruction, temps, stores, placeholders);
+                }
             }
 
-            // assign from rhs values to lhs receivers
-            int numAssignments = node.Assignments.Length;
+            int numAssignments = node.AssignmentSteps.Length;
             for (int i = 0; i < numAssignments; i++)
             {
-                // lower the assignment and replace the placeholders for source and target in the process
-                var assignmentInfo = node.Assignments[i];
-
-                AddPlaceholderReplacement(assignmentInfo.LValuePlaceholder, lhsReceivers[i]);
-                AddPlaceholderReplacement(assignmentInfo.RValuePlaceholder, rhsValues[i]);
+                // lower the assignment and replace the placeholders for its outputs in the process
+                var assignmentInfo = node.AssignmentSteps[i];
+                AddPlaceholderReplacement(assignmentInfo.OutputPlaceholder, lhsTemps[i]);
 
                 var assignment = VisitExpression(assignmentInfo.Assignment);
 
-                RemovePlaceholderReplacement(assignmentInfo.LValuePlaceholder);
-                RemovePlaceholderReplacement(assignmentInfo.RValuePlaceholder);
+                RemovePlaceholderReplacement(assignmentInfo.OutputPlaceholder);
 
                 stores.Add(assignment);
             }
 
-            stores.Add(new BoundVoid(node.Syntax, node.Type));
-            var result = _factory.Sequence(temps.ToImmutable(), stores.ToArray());
+            var result = _factory.Sequence(temps.ToImmutable(), stores.ToImmutable(), new BoundVoid(node.Syntax, node.Type));
+
+            RemovePlaceholderReplacements(placeholders);
+            placeholders.Free();
 
             temps.Free();
             stores.Free();
-            lhsReceivers.Free();
 
             return result;
         }
 
-        private ImmutableArray<BoundExpression> AccessTupleFields(BoundDeconstructionAssignmentOperator node, BoundExpression loweredRight, ArrayBuilder<LocalSymbol> temps, ArrayBuilder<BoundExpression> stores)
+        /// <summary>
+        /// Adds the side effects to stores and returns temporaries (as a flat list) to access them.
+        /// </summary>
+        private ImmutableArray<BoundExpression> LhsSideEffects(ImmutableArray<BoundExpression> variables, ArrayBuilder<LocalSymbol> temps, ArrayBuilder<BoundExpression> stores)
+        {
+            var lhsReceivers = ArrayBuilder<BoundExpression>.GetInstance(variables.Length);
+
+            foreach (var variable in variables)
+            {
+                lhsReceivers.Add(TransformCompoundAssignmentLHS(variable, stores, temps, isDynamicAssignment: false));
```

Comment:
Not related to this change, but are we certain that isDynamicAssignment should always be false?